src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx

```diff
@@ -25,61 +25,102 @@ export function SampleMetadataImportSelectFile(): JSX.Element {
   const dispatch: ImportDispatch = useImportDispatch();
   const [status, setStatus] = React.useState<StepsProps["status"]>("process");
   const [loading, setLoading] = React.useState<boolean>(false);
+  const [valid, setValid] = React.useState<boolean>(true);
 
   React.useEffect(() => {
     if (projectId != null) {
       dispatch(setProjectId(projectId));
     }
   }, [dispatch, projectId]);
 
+  const readFileContents = (file: Blob) => {
+    const reader = new FileReader();
+
+    return new Promise((resolve, reject) => {
+      reader.onerror = () => {
+        reader.abort();
+        reject("Problem reading the file.");
+      };
+
+      reader.onload = () => {
+        resolve(reader.result);
+      };
+
+      reader.readAsBinaryString(file);
+    });
+  };
+
   const options = {
     multiple: false,
     showUploadList: false,
     accept: ".xls,.xlsx,.csv",
-    beforeUpload: (file: RcFile) => {
+    customRequest: () => {
+      navigate(`/${projectId}/sample-metadata/upload/columns`);
+    },
+    beforeUpload: async (file: RcFile) => {
       try {
-        setLoading(true);
-        const reader = new FileReader();
-        if (reader.readAsBinaryString) {
-          reader.onload = async () => {
-            const workbook: WorkBook = XLSX.read(reader.result, {
-              type: "binary",
-              raw: true,
-            });
-            const { SheetNames } = workbook;
-            const [firstSheet] = SheetNames;
-            const rows: MetadataItem[] = XLSX.utils.sheet_to_json(
-              workbook.Sheets[firstSheet],
-              {
-                rawNumbers: false,
-              }
-            );
-            const cleanRows = JSON.parse(
-              JSON.stringify(rows).replace(/"\s+|\s+"/g, '"')
-            );
-            await dispatch(setHeaders({ headers: Object.keys(cleanRows[0]) }));
-            await dispatch(setMetadata(cleanRows));
-          };
-          reader.readAsBinaryString(file);
+        const data = await readFileContents(file);
+        const workbook: XLSX.WorkBook = XLSX.read(data, {
+          type: "binary",
+          raw: true,
+        });
+        const { SheetNames } = workbook;
+        const [firstSheet] = SheetNames;
+        // Not loving how I'm parsing the file twice to get the raw headers
+        // See https://docs.sheetjs.com/docs/api/utilities
+        const headers: string[] = XLSX.utils.sheet_to_json(
```

Comment:
As discussed, update this to only convert the sheet to json once

Reply:
Tried doing what we discussed, but it didn't work because when
`const rows: MetadataItem[] = XLSX.utils.sheet_to_json(workbook.Sheets[firstSheet], {rawNumbers: false,});`
is called, the excel document is stored in a key-value hash. Key-value hashes cannot have duplicate header names, their keys must be unique. This function automatically alters the headers names making them unique and then I cannot find unique names later.

Reply:
Thanks for updating this! :)